Null pointer arithmetic warning building under Xcode/macOS

Source lines (132)

https://github.com/cgreen-devs/cgreen/blob/d06fb7e9800bc7fa7098fcf7bc421076f4b870ed/tools/gopt.c#L132-L140

Compiler output:

```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for include file pthread.h
-- Looking for include file pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Performing Test WITH_FPIC
-- Performing Test WITH_FPIC - Success
-- Performing Test WITH_FORTIFY_SOURCE
-- Performing Test WITH_FORTIFY_SOURCE - Success
-- Looking for sys/resource.h
-- Looking for sys/resource.h - found
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/dmeehan/Documents/Dropbox/Projects/Wavetrend/RX14XX Firmware/RX14XX-API/lib/cgreen/build
...
[ 21%] Building C object tools/CMakeFiles/cgreen-runner.dir/gopt.c.o
/Users/dmeehan/Documents/Dropbox/Projects/Wavetrend/RX14XX Firmware/RX14XX-API/lib/cgreen/tools/gopt.c:132:53: warning: arithmetic on a null pointer
      treated as a cast from integer to pointer is a GNU extension [-Wnull-pointer-arithmetic]
                                        if((char*)0 + 1 == next_option-> arg) {
                                           ~~~~~~~~ ^
1 warning generated.
...
```

--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -1,7 +1,7 @@
 include_directories(${CGREEN_PUBLIC_INCLUDE_DIRS} ${PROJECT_BINARY_DIR})
 
 set(RUNNER_SRCS
-    cgreen-runner.c gopt.c runner.c discoverer.c test_item.c io.c)
+    cgreen-runner.c gopt.c gopt-errors.c runner.c discoverer.c test_item.c io.c)
 set_source_files_properties(${RUNNER_SRCS} PROPERTIES LANGUAGE C)
 
 add_executable(cgreen-runner ${RUNNER_SRCS})

--- a/tools/cgreen-runner.c
+++ b/tools/cgreen-runner.c
@@ -30,9 +30,9 @@ static void version(void) {
 
 
 /*----------------------------------------------------------------------*/
-static void usage(const char **argv) {
+static void usage(const char *program_name) {
     printf("cgreen-runner for Cgreen unittest and mocking framework v%s\n\n", VERSION);
-    printf("Usage:\n    %s [--xml <prefix>] [--suite <name>] [--verbose] [--quiet] [--no-run] [--help] (<library> [<test>])+\n\n", argv[0]);
+    printf("Usage:\n    %s [--xml <prefix>] [--suite <name>] [--verbose] [--quiet] [--no-run] [--help] (<library> [<test>])+\n\n", program_name);
     printf("Discover and run all or named cgreen test(s) from one or multiple\n");
     printf("dynamically loadable libraries.\n\n");
     printf("A single test can be run using the form [<context>:]<name> where <context> can\n");
@@ -52,14 +52,87 @@ static void usage(const char **argv) {
 
 /*======================================================================*/
 
-static void *options = NULL;
+enum option_arg {
+    HELP_OPT,
+    XML_OPT,
+    SUITE_OPT,
+    VERBOSE_OPT,
+    QUITE_OPT,
+    VERSION_OPT,
+    COLORS_OPT,
+    COLOURS_OPT,
+    NO_COLORS_OPT,
+    NO_COLOURS_OPT,
+    NO_RUN_OPT,
+    LAST_OPT,
+};
+
+static struct option options[] = {
+    [HELP_OPT] = {
+        .short_name = 'h',
+        .long_name = "help",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [XML_OPT] = {
+        .short_name = 'x',
+        .long_name = "xml",
+        .flags = GOPT_ARGUMENT_REQUIRED,
+    },
+    [SUITE_OPT] = {
+        .short_name = 's',
+        .long_name = "suite",
+        .flags = GOPT_ARGUMENT_REQUIRED,
+    },
+    [VERBOSE_OPT] = {
+        .short_name = 'v',
+        .long_name = "verbose",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [QUITE_OPT] = {
+        .short_name = 'q',
+        .long_name = "quite",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [VERSION_OPT] = {
+        .short_name = 'V',
+        .long_name = "version",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [COLORS_OPT] = {
+        .short_name = 'c',
+        .long_name = "colors",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [COLOURS_OPT] = {
+        .short_name = 'c',
+        .long_name = "colours",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [NO_COLORS_OPT] = {
+        .short_name = 'C',
+        .long_name = "no-colors",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [NO_COLOURS_OPT] = {
+        .short_name = 'C',
+        .long_name = "no-colours",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [NO_RUN_OPT] = {
+        .short_name = 'n',
+        .long_name = "no-run",
+        .flags = GOPT_ARGUMENT_FORBIDDEN,
+    },
+    [LAST_OPT] = {
+        .flags = GOPT_LAST,
+    },
+};
 static TestReporter *reporter = NULL;
 static TextReporterOptions reporter_options;
 
 static void cleanup(void)
 {
     if (reporter) reporter->destroy(reporter);
-    if (options) gopt_free(options);
 }
 
 static char* get_a_suite_name(const char *suite_option, const char *test_library_name) {
@@ -79,76 +152,9 @@ static char* get_a_suite_name(const char *suite_option, const char *test_library
     }
 }
 
-
-/*----------------------------------------------------------------------*/
-static int initialize_option_handling(int argc, const char **argv) {
-    options = gopt_sort(&argc, argv, gopt_start(
-                                                gopt_option('x',
-                                                            GOPT_ARG,
-                                                            gopt_shorts('x'),
-                                                            gopt_longs("xml")
-                                                            ),
-                                                gopt_option('s',
-                                                            GOPT_ARG,
-                                                            gopt_shorts('s'),
-                                                            gopt_longs("suite")
-                                                            ),
-                                                gopt_option('v',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('v'),
-                                                            gopt_longs("verbose")
-                                                            ),
-                                                gopt_option('V',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('V'),
-                                                            gopt_longs("version")
-                                                            ),
-                                                gopt_option('c',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('c'),
-                                                            gopt_longs("colours")
-                                                            ),
-                                                gopt_option('c',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('c'),
-                                                            gopt_longs("colors")
-                                                            ),
-                                                gopt_option('C',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('C'),
-                                                            gopt_longs("no-colors")
-                                                            ),
-                                                gopt_option('C',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('C'),
-                                                            gopt_longs("no-colours")
-                                                            ),
-                                                gopt_option('q',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('q'),
-                                                            gopt_longs("quiet")
-                                                            ),
-                                                gopt_option('n',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('n'),
-                                                            gopt_longs("no-run")
-                                                            ),
-                                                gopt_option('h',
-                                                            GOPT_NOARG,
-                                                            gopt_shorts('h'),
-                                                            gopt_longs("help")
-                                                            )
-                                                )
-                        );
-    return(argc);
-}
-
-
 /*----------------------------------------------------------------------*/
 static bool have_xml_option(void) {
-    const char *prefix_option;
-
-    return gopt_arg(options, 'x', &prefix_option);
+    return options[XML_OPT].count > 0;
 }
 
 /*----------------------------------------------------------------------*/
@@ -195,62 +201,55 @@ static void inhibit_appropriate_suite_message(int i, int library_count) {
 
 
 /*======================================================================*/
-int main(int argc, const char **argv) {
+int main(int argc, char **argv) {
     int i;
 
     bool verbose = false;
     bool no_run = false;
 
-    const char *prefix_option;
     const char *suite_name_option = NULL;
-    const char *tmp;
 
     bool any_fail = false;
 
     atexit(cleanup);
 
-    argc = initialize_option_handling(argc, argv);
+    argc = gopt(argv, options);
+    gopt_errors(argv[0], options);
 
-    if (gopt_arg(options, 'x', &prefix_option))
-        reporter = create_xml_reporter(prefix_option);
+    if (options[XML_OPT].count)
+        reporter = create_xml_reporter(options[XML_OPT].argument);
     else
         reporter = create_text_reporter();
 
-    gopt_arg(options, 's', &suite_name_option);
+    suite_name_option = options[SUITE_OPT].argument;
 
-    if (gopt_arg(options, 'v', &tmp))
-        verbose = true;
-
-    if (gopt_arg(options, 'n', &tmp))
-        no_run = true;
+    verbose = options[VERBOSE_OPT].count > 0;
+    no_run = options[NO_RUN_OPT].count > 0;
 
     reporter_options.use_colours = true;
     if (isatty(fileno(stdout))) {
-        if (gopt_arg(options, 'C', &tmp))
+        if (options[NO_COLORS_OPT].count || options[NO_COLOURS_OPT].count)
             reporter_options.use_colours = false;
     } else {
-        if (!gopt_arg(options, 'c', &tmp))
+        if (!options[COLORS_OPT].count && !options[COLOURS_OPT].count)
             reporter_options.use_colours = false;
     }
     panic_use_colours = reporter_options.use_colours;
 
-    if (gopt_arg(options, 'q', &tmp))
-        reporter_options.quiet_mode = true;
-    else
-        reporter_options.quiet_mode = false;
+    reporter_options.quiet_mode = options[QUITE_OPT].count > 0;
 
-    if (gopt_arg(options, 'h', &tmp)) {
-        usage(argv);
+    if (options[HELP_OPT].count) {
+        usage(argv[0]);
         return EXIT_SUCCESS;
     }
 
-    if (gopt_arg(options, 'V', &tmp)) {
+    if (options[VERSION_OPT].count) {
         version();
         return EXIT_SUCCESS;
     }
 
     if (argc < 2) {
-        usage(argv);
+        usage(argv[0]);
         return EXIT_FAILURE;
     }
 

--- a/tools/gopt-errors.c
+++ b/tools/gopt-errors.c
@@ -1,0 +1,86 @@
+/* gopt-errors.c   PUBILC DOMAIN 2015   t.gopt@purposeful.co.uk */
+
+/* <http:///www.purposeful.co.uk/software/gopt> */
+
+/*
+  I, Tom Vajzovic, am the author of this software and its documentation.
+  I permanently abandon all intellectual property rights in them, including
+  copyright, trademarks, design rights, database right, patents, and the right
+  to be identified as the author.
+
+  I am fairly certain that the software does what the documentation says it
+  does, but I do not guarantee that it does, or that it does what you think it
+  should.  I do not guarantee that it will not have undesirable side effects.
+
+  You are free to use, modify and distribute this software as you please, but
+  you do so at your own risk.  If you do not pass on this warning then you may
+  be responsible for any problems encountered by those who obtain the software
+  through you.
+*/
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#ifdef _BSD_SOURCE
+#include <sysexits.h>
+#else
+#define EX_USAGE EXIT_FAILURE
+#endif
+
+#include "gopt.h"
+
+void gopt_errors (const char *argv0, const struct option *options)
+{
+  unsigned int bad_option_index = 0;
+  unsigned int i;
+
+  while (!(options[bad_option_index].flags & GOPT_LAST))
+  {
+    bad_option_index++;
+  }
+
+  if (options[bad_option_index].short_name)
+  {
+    fprintf (stderr, "%s: unrecognised option -%c\n", argv0, options[bad_option_index].short_name);
+    exit (EX_USAGE);
+  }
+
+  if (options[bad_option_index].long_name)
+  {
+    fprintf (stderr, "%s: unrecognised option --%s\n", argv0, options[bad_option_index].long_name);
+    exit (EX_USAGE);
+  }
+
+  for (i = 0; i < bad_option_index; i++)
+  {
+    if ((options[i].flags & GOPT_SEEN_SHORT_WITHOUT) && (options[i].flags & GOPT_ARGUMENT_REQUIRED))
+    {
+      fprintf (stderr, "%s: option -%c requires an option-argument\n", argv0, options[i].short_name);
+      exit (EX_USAGE);
+    }
+
+    if ((options[i].flags & GOPT_SEEN_SHORT_WITH) && (options[i].flags & GOPT_ARGUMENT_FORBIDDEN))
+    {
+      fprintf (stderr, "%s: option -%c must not have an option-argument\n", argv0, options[i].short_name);
+      exit (EX_USAGE);
+    }
+
+    if ((options[i].flags & GOPT_SEEN_LONG_WITHOUT) && (options[i].flags & GOPT_ARGUMENT_REQUIRED))
+    {
+      fprintf (stderr, "%s: option --%s requires an option-argument\n", argv0, options[i].long_name);
+      exit (EX_USAGE);
+    }
+
+    if ((options[i].flags & GOPT_SEEN_LONG_WITH) && (options[i].flags & GOPT_ARGUMENT_FORBIDDEN))
+    {
+      fprintf (stderr, "%s: option --%s must not have an option-argument\n", argv0, options[i].long_name);
+      exit (EX_USAGE);
+    }
+
+    if ((options[i].count > 1) && !(options[i].flags & GOPT_REPEATABLE))
+    {
+      fprintf (stderr, "%s: option -%c/--%s may not be repeated\n", argv0, options[i].short_name, options[i].long_name);
+      exit (EX_USAGE);
+    }
+  }
+}

--- a/tools/gopt.c
+++ b/tools/gopt.c
@@ -1,266 +1,226 @@
-/* gopt.c version 8.1: tom.viza@gmail.com PUBLIC DOMAIN 2003-8 */
-/*
-  I, Tom Vajzovic, am the author of this software and its documentation and
-  permanently abandon all copyright and other intellectual property rights in
-  them, including the right to be identified as the author.
+/* gopt.c   PUBILC DOMAIN 2015   t.gopt@purposeful.co.uk */
 
-  I am fairly certain that this software does what the documentation says it
-  does, but I cannot guarantee that it does, or that it does what you think it
-  should, and I cannot guarantee that it will not have undesirable side effects.
+/* <http:///www.purposeful.co.uk/software/gopt> */
+
+/*
+  I, Tom Vajzovic, am the author of this software and its documentation.
+  I permanently abandon all intellectual property rights in them, including
+  copyright, trademarks, design rights, database right, patents, and the right
+  to be identified as the author.
+
+  I am fairly certain that the software does what the documentation says it
+  does, but I do not guarantee that it does, or that it does what you think it
+  should.  I do not guarantee that it will not have undesirable side effects.
 
   You are free to use, modify and distribute this software as you please, but
-  you do so at your own risk.  If you remove or hide this warning then you are
-  responsible for any problems encountered by people that you make the software
-  available to.
-
-  Before modifying or distributing this software I ask that you would please
-  read http://www.purposeful.co.uk/tfl/
+  you do so at your own risk.  If you do not pass on this warning then you may
+  be responsible for any problems encountered by those who obtain the software
+  through you.
 */
 
-#include <stdio.h>
-#include <stdlib.h>
 #include <string.h>
-#include "gopt.h"
 
-#ifdef USE_SYSEXITS
-#include <sysexits.h>
+#ifdef _WIN32
+#define strncasecmp _strnicmp
 #else
-#define EX_OSERR EXIT_FAILURE
-#define EX_USAGE EXIT_FAILURE
+#include <strings.h>
 #endif
 
-struct opt_spec_s {
-    int key;
-    int flags;
-    const char *shorts;
-    const char* const *longs;
-};
-typedef struct opt_spec_s opt_spec_t;
+#include "gopt.h"
 
-struct opt_s {
-    int key;
-    const char *arg;
-};
-typedef struct opt_s opt_t;
 
-void *gopt_sort(int *argc, const char **argv, const void *opt_specs) {
-    void *opts;
-    {{{
-                const char* const *arg_p= argv + 1;
-                size_t opt_count= 1;
-                for(; *arg_p; ++arg_p)
-                    if('-' == (*arg_p)[0] && (*arg_p)[1]) {
-                        if('-' == (*arg_p)[1])
-                            if((*arg_p)[2])
-                                ++opt_count;
-                            else
-                                break;
-                        else {
-                            const opt_spec_t *opt_spec_p= opt_specs;
-                            for(; opt_spec_p-> key; ++opt_spec_p)
-                                if(strchr(opt_spec_p-> shorts, (*arg_p)[1])) {
-                                    opt_count+= opt_spec_p-> flags & GOPT_ARG ? 1 : strlen((*arg_p) + 1);
-                                    break;
-                                }
-                        }
-                    }
-                opts= malloc(opt_count * sizeof(opt_t));
-            }}}
+/* returns index of first exact match or only prefix match, or last index */
+static unsigned int long_option_get_index (const char *arg, const struct option *options)
+{
+  unsigned int count = 0;
+  unsigned int found;
+  unsigned int i;
+
+  size_t arg_len = strcspn (arg, "=");
+
+  for (i = 0; !(options[i].flags & GOPT_LAST); i++)
+  {
+    if (options[i].long_name)
     {
-        const char **arg_p= argv + 1;
-        const char **next_operand= arg_p;
-        opt_t *next_option= opts;
-      
-        if(! opts) {
-            perror(argv[0]);
-            exit(EX_OSERR);
-        }  
-        for(; *arg_p; ++arg_p)
-            if('-' == (*arg_p)[0] && (*arg_p)[1])
-                if('-' == (*arg_p)[1])
-                    if((*arg_p)[2])
-                        {{{
-                                    const opt_spec_t *opt_spec_p= opt_specs;
-                                    const char* const *longs= opt_spec_p-> longs;
-                                    next_option-> key= 0;
-                                    while(*longs) {
-                                        const char *option_cp= (*arg_p) + 2;
-                                        const char *name_cp= *longs;
-                                        while(*option_cp && *option_cp == *name_cp) {
-                                            ++option_cp;
-                                            ++name_cp;
-                                        }
-                                        if('=' == *option_cp || !*option_cp) {
-                                            if(*name_cp) {
-                                                if(next_option-> key) {
-                                                    fprintf(stderr, "%s: --%.*s: abbreviated option is ambiguous\n", argv[0], (int)(option_cp -((*arg_p) + 2)), (*arg_p) + 2);
-                                                    free(opts);
-                                                    exit(EX_USAGE);
-                                                }
-                                                next_option-> key= opt_spec_p-> key;
-                                            }
-                                            else {
-                                                next_option-> key= opt_spec_p-> key;
-                                                goto found_long;
-                                            }
-                                        }
-                                        if(!*++longs) {
-                                            ++opt_spec_p;
-                                            if(opt_spec_p-> key)
-                                                longs= opt_spec_p-> longs;
-                                        }
-                                    }
-                                    if(! next_option-> key) {
-                                        fprintf(stderr, "%s: --%.*s: unknown option\n", argv[0], (int)strcspn((*arg_p) + 2, "="), (*arg_p) + 2);
-                                        free(opts);
-                                        exit(EX_USAGE);              
-                                    }
-                                    for(opt_spec_p= opt_specs; opt_spec_p-> key != next_option-> key; ++opt_spec_p);
-                                found_long:
-            
-                                    if(!(opt_spec_p-> flags & GOPT_REPEAT)) {
-                                        const opt_t *opt_p= opts;
-                                        for(; opt_p != next_option; ++opt_p)
-                                            if(opt_p-> key == opt_spec_p-> key) {
-                                                fprintf(stderr, "%s: --%.*s: option may not be repeated (in any long or short form)\n", argv[0], (int)strcspn((*arg_p) + 2, "="), (*arg_p) + 2);
-                                                free(opts);
-                                                exit(EX_USAGE);
-                                            }
-                                    }
-                                    if(opt_spec_p-> flags & GOPT_ARG) {
-                                        next_option-> arg= strchr((*arg_p) + 2, '=') + 1;
-                                        if((char*)0 + 1 == next_option-> arg) {
-                                            ++arg_p;
-                                            if(!*arg_p || ('-' == (*arg_p)[0] && (*arg_p)[1])) {
-                                                fprintf(stderr, "%s: --%s: option requires an option argument\n", argv[0], (*(arg_p-1)) + 2);
-                                                free(opts);
-                                                exit(EX_USAGE);
-                                            }
-                                            next_option-> arg= *arg_p;
-                                        }
-                                    }
-                                    else {
-                                        if(strchr((*arg_p) + 2, '=')) {
-                                            fprintf(stderr, "%s: --%.*s: option may not take an option argument\n", argv[0], (int)strcspn((*arg_p) + 2, "="), (*arg_p) + 2);
-                                            free(opts);
-                                            exit(EX_USAGE);
-                                        }
-                                        next_option-> arg= NULL;
-                                    }
-                                    ++next_option;
-                                }}}
-                    else {
-                        for(++arg_p; *arg_p; ++arg_p)
-                            *next_operand++= *arg_p;
-                        break;
-                    }
-                else
-                    {{{
-                                const char *short_opt= (*arg_p) + 1;
-                                for(;*short_opt; ++short_opt) {
-                                    const opt_spec_t *opt_spec_p= opt_specs;
-            
-                                    for(; opt_spec_p-> key; ++opt_spec_p)
-                                        if(strchr(opt_spec_p-> shorts, *short_opt)) {
-                                            if(!(opt_spec_p-> flags & GOPT_REPEAT)) {
-                                                const opt_t *opt_p= opts;
-                                                for(; opt_p != next_option; ++opt_p)
-                                                    if(opt_p-> key == opt_spec_p-> key) {
-                                                        fprintf(stderr, "%s: -%c: option may not be repeated (in any long or short form)\n", argv[0], *short_opt);
-                                                        free(opts);
-                                                        exit(EX_USAGE);
-                                                    }
-                                            }
-                                            next_option-> key= opt_spec_p-> key;
+      size_t full_len = strlen (options[i].long_name);
 
-                                            if(opt_spec_p-> flags & GOPT_ARG) {
-                                                if(short_opt[1])
-                                                    next_option-> arg= short_opt + 1;
-                  
-                                                else {
-                                                    ++arg_p;
-                                                    if(!*arg_p || ('-' == (*arg_p)[0] && (*arg_p)[1])) {
-                                                        fprintf(stderr, "%s: -%c: option requires an option argument\n", argv[0], *short_opt);
-                                                        free(opts);
-                                                        exit(EX_USAGE);
-                                                    }
-                                                    next_option-> arg= *arg_p;
-                                                }
-                                                ++next_option;
-                                                goto break_2;
-                                            }
-                                            next_option-> arg= NULL;
-                                            ++next_option;
-                                            goto continue_2;
-                                        }
-                                    fprintf(stderr, "%s: -%c: unknown option\n", argv[0], *short_opt);
-                                    free(opts);
-                                    exit(EX_USAGE);
-                                continue_2: ;
-                                }
-                            break_2: ;
-                            }}}
-            else
-                *next_operand++= *arg_p;
+      if ((arg_len <= full_len) && !strncasecmp (options[i].long_name, arg, arg_len))
+      {
+        if (arg_len == full_len)
+        {
+          return i;
+        }
 
-        next_option-> key= 0;
-        *next_operand= NULL;
-        *argc= next_operand - argv;
+        found = i;
+        count++;
+      }
     }
-    return opts;
+  }
+
+  return ((count == 1) ? found : i);
 }
 
-size_t gopt(const void *vptr_opts, int key) {
-    const opt_t *opts= vptr_opts;
-    size_t count= 0;
-    for(; opts-> key; ++opts)
-        count+= opts-> key == key;
+/* returns index of first match, or last index */
+static unsigned int short_option_get_index (char c, const struct option *options)
+{
+  unsigned int i;
 
-    return count;
+  for (i = 0; !(options[i].flags & GOPT_LAST); i++)
+  {
+    if (options[i].short_name == c)
+    {
+      return i;
+    }
+  }
+
+  return i;
 }
 
-size_t gopt_arg(const void *vptr_opts, int key, const char **arg) {
-    const opt_t *opts= vptr_opts;
-    size_t count= 0;
- 
-    for(; opts-> key; ++opts)
-        if(opts-> key == key) {
-            if(! count)
-                *arg= opts-> arg;
-            ++count;
+int gopt (char **argv, struct option *options)
+{
+  unsigned int operand_count = 1;
+  unsigned int doubledash    = 0;
+  unsigned int expecting     = 0;
+  unsigned int option_index;
+  unsigned int i, j;
+
+  for (i = 0; !(options[i].flags & GOPT_LAST); i++)
+  {
+    options[i].argument = NULL;
+    options[i].count    = 0;
+  }
+
+  options[i].short_name = 0;
+  options[i].long_name  = NULL;
+  options[i].argument   = NULL;
+  options[i].count      = 0;
+
+  for (i = 1; argv[i]; i++)
+  {
+    if (doubledash)
+    {
+      argv[operand_count] = argv[i];
+      operand_count++;
+      continue;
+    }
+
+    if (expecting)
+    {
+      if ((argv[i][0] != '-') || !argv[i][1] || !(options[option_index].flags & GOPT_ARGUMENT_NO_HYPHEN))
+      {
+        options[option_index].flags    |= expecting;
+        options[option_index].argument  = argv[i];
+        expecting = 0;
+        continue;
+      }
+      else
+      {
+        options[option_index].flags    |= (expecting >> 1); /* change _WITH to _WITHOUT */
+        options[option_index].argument  =  NULL;
+        expecting = 0;
+      }
+    }
+
+    if ((argv[i][0] == '-') && (argv[i][1] == '-') && (argv[i][2] == 0))
+    {
+      doubledash = 1;
+      continue;
+    }
+
+    if ((argv[i][0] == '-') && (argv[i][1] == '-'))
+    {
+      char *eq = strchr (&argv[i][2], '=');
+
+      option_index = long_option_get_index (&argv[i][2], options);
+
+      options[option_index].count++;
+
+      if ((options[option_index].flags & GOPT_LAST) && !options[option_index].long_name)
+      {
+        options[option_index].long_name = &argv[i][2];
+      }
+
+      if (eq)
+      {
+        options[option_index].argument  = (eq + 1);
+        options[option_index].flags    |= GOPT_SEEN_LONG_WITH;
+      }
+      else if (options[option_index].flags & GOPT_ARGUMENT_REQUIRED)
+      {
+        expecting = GOPT_SEEN_LONG_WITH;
+      }
+      else
+      {
+        options[option_index].argument  = NULL;
+        options[option_index].flags    |= GOPT_SEEN_LONG_WITHOUT;
+      }
+    }
+    else if ((argv[i][0] == '-') && argv[i][1])
+    {
+      for (j = 1; argv[i][j]; j++)
+      {
+        option_index = short_option_get_index (argv[i][j], options);
+
+        options[option_index].count++;
+
+        if (options[option_index].flags & GOPT_LAST)
+        {
+          if (!options[option_index].short_name)
+          {
+            options[option_index].short_name = argv[i][j];
+          }
+
+          if (argv[i][j+1])
+          {
+            options[option_index].argument  = &argv[i][j+1];
+            options[option_index].flags    |= GOPT_SEEN_SHORT_WITH;
+          }
+          else
+          {
+            options[option_index].argument  = NULL;
+            options[option_index].flags    |= GOPT_SEEN_SHORT_WITHOUT;
+          }
+          break;
         }
-    return count;
-}
 
-const char *gopt_arg_i(const void *vptr_opts, int key, size_t i) {
-    const opt_t *opts= vptr_opts;
-  
-    for(; opts-> key; ++opts)
-        if(opts-> key == key) {
-            if(! i)
-                return opts-> arg;      
-            --i;
+        if (options[option_index].flags & GOPT_ARGUMENT_FORBIDDEN)
+        {
+          options[option_index].argument  = NULL;
+          options[option_index].flags    |= GOPT_SEEN_SHORT_WITHOUT;
         }
-    return NULL;
-}
 
-size_t gopt_args(const void *vptr_opts, int key, const char **args, size_t args_len) {
-    const char **args_stop= args + args_len;
-    const char **args_ptr= args;
-    const opt_t *opts= vptr_opts;
-
-    for(; opts-> key; ++opts)
-        if(opts-> key == key) {
-            if(args_stop == args_ptr)
-                return args_len + gopt(opts, key);
-      
-            *args_ptr++= opts-> arg;
+        else if (argv[i][j+1])
+        {
+          options[option_index].argument  = &argv[i][j+1];
+          options[option_index].flags    |= GOPT_SEEN_SHORT_WITH;
+          break;
         }
-    if(args_stop != args_ptr)
-        *args_ptr= NULL;
-    return args_ptr - args;
-}
 
-void gopt_free(void *vptr_opts) {
-    free(vptr_opts);
+        else if (options[option_index].flags & GOPT_ARGUMENT_REQUIRED)
+        {
+          expecting = GOPT_SEEN_SHORT_WITH;
+        }
+
+        else
+        {
+          options[option_index].argument  = NULL;
+          options[option_index].flags    |= GOPT_SEEN_SHORT_WITHOUT;
+        }
+      }
+    }
+    else
+    {
+      argv[operand_count] = argv[i];
+      operand_count++;
+    }
+  }
+
+  if (expecting)
+  {
+    options[option_index].flags    |= (expecting >> 1); /* change _WITH to _WITHOUT */
+    options[option_index].argument  =  NULL;
+  }
+
+  argv[operand_count] = NULL;
+  return operand_count;
 }

--- a/tools/gopt.h
+++ b/tools/gopt.h
@@ -1,62 +1,143 @@
-/* gopt.h version 8.1: tom.viza@gmail.com PUBLIC DOMAIN 2003-8 */
+/* gopt.h   PUBILC DOMAIN 2015   t.gopt@purposeful.co.uk */
+
+/* <http:///www.purposeful.co.uk/software/gopt> */
+
 /*
-I, Tom Vajzovic, am the author of this software and its documentation and
-permanently abandon all copyright and other intellectual property rights in
-them, including the right to be identified as the author.
+  I, Tom Vajzovic, am the author of this software and its documentation.
+  I permanently abandon all intellectual property rights in them, including
+  copyright, trademarks, design rights, database right, patents, and the right
+  to be identified as the author.
 
-I am fairly certain that this software does what the documentation says it
-does, but I cannot guarantee that it does, or that it does what you think it
-should, and I cannot guarantee that it will not have undesirable side effects.
+  I am fairly certain that the software does what the documentation says it
+  does, but I do not guarantee that it does, or that it does what you think it
+  should.  I do not guarantee that it will not have undesirable side effects.
 
-You are free to use, modify and distribute this software as you please, but
-you do so at your own risk.  If you remove or hide this warning then you are
-responsible for any problems encountered by people that you make the software
-available to.
-
-Before modifying or distributing this software I ask that you would please
-read http://www.purposeful.co.uk/tfl/
+  You are free to use, modify and distribute this software as you please, but
+  you do so at your own risk.  If you do not pass on this warning then you may
+  be responsible for any problems encountered by those who obtain the software
+  through you.
 */
 
 #ifndef GOPT_H_INCLUDED
 #define GOPT_H_INCLUDED
 
-#define GOPT_ONCE   0
-#define GOPT_REPEAT 1
-#define GOPT_NOARG  0
-#define GOPT_ARG    2
+#ifdef __cplusplus
+extern "C" {
+#endif
 
-struct gopt_option_spec { int key; int flags; const char *shortname; const char*const*longname; };
+struct option
+{
+  /* input: */
+  char          short_name;
+  const char   *long_name;
 
-#define gopt_start(...)  (const void*)(const struct gopt_option_spec []){ __VA_ARGS__, {0, 0, 0, 0}}
-#define gopt_option(key,flags,shortname,longname)    { key, flags, shortname, longname }
-#define gopt_shorts( ... )      (const char*)(const char[]){ __VA_ARGS__, 0 }
-#define gopt_longs( ... )       (const char**)(const char*[]){ __VA_ARGS__, NULL }
+  /* input/output: */
+  unsigned int  flags;
 
+  /* output: */
+  unsigned int  count;
+  char         *argument;
+};
 
-void *gopt_sort( int *argc, const char **argv, const void *opt_specs );
-/* returns a pointer for use in the following calls
- * prints to stderr and call exit() on error
- */
-size_t gopt( const void *opts, int key );
-/* returns the number of times the option was specified
- * which will be 0 or 1 unless GOPT_REPEAT was used
- */
-size_t gopt_arg( const void *opts, int key, const char **arg );
-/* returns the number of times the option was specified
- * writes a pointer to the option argument from the first (or only) occurance to *arg
- */
-const char *gopt_arg_i( const void *opts, int key, size_t i );
-/* returns a pointer to the ith (starting at zero) occurance
- * of the option, or NULL if it was not specified that many times
- */
-size_t gopt_args( const void *opts, int key, const char **args, size_t args_len );
-/* returns the number of times the option was specified
- * writes pointers to the option arguments in the order of occurance to args[].
- * writes at most args_len pointers
- * if the return value is less than args_len, also writes a null pointer
- */
-void gopt_free( void *opts );
-/* releases memory allocated in the corresponding call to gopt_sort()
- * opts can no longer be used 
- */
+/* values for flags: */
+#define GOPT_ARGUMENT_OPTIONAL  0x000u /* option may or may not have an option-argument */
+#define GOPT_ARGUMENT_FORBIDDEN 0x001u /* option must not have an option-argument */
+#define GOPT_ARGUMENT_REQUIRED  0x002u /* option must have an option-argument */
+#define GOPT_ARGUMENT_NO_HYPHEN 0x004u /* option-argument may not start with hyphen/minus */
+#define GOPT_REPEATABLE         0x008u /* option may be specified more than once */
+#define GOPT_SEEN_SHORT_WITHOUT 0x010u /* short form of option was present without an option-argument in argv */
+#define GOPT_SEEN_SHORT_WITH    0x020u /* short form of option was present with an option-argument in argv */
+#define GOPT_SEEN_LONG_WITHOUT  0x040u /* long form of option was present without an option-argument in argv */
+#define GOPT_SEEN_LONG_WITH     0x080u /* long form of option was present with an option-argument in argv */
+#define GOPT_LAST               0x100u /* this is the last element of the array */
+
+int gopt (char **argv, struct option *options);
+/*
+  The function gopt() takes an argument vector argv, which has the same
+  form as the second argument to main().  It removes from the vector all
+  options and option-arguments, leaving only the program name, the
+  operands and a null pointer at the end.  It returns the updated
+  argument count argc.  It does not need to know the initial value of
+  argc because the input vector must be null-terminated.
+
+  It also takes an array of struct option, the elements of which are
+  used for specifying the options which are to be recognized and for
+  returning details of which options were present in the argument
+  vector.
+
+  The members of struct option are used as follows:
+
+  On input, short_name is the single character that follows "-" to make
+  a short option, or zero if there is no short form of this option.
+
+  On input, long_name is the name of the long option excluding the "--",
+  or NULL if there is no long form of this option.
+
+  On return short_name and long_name are unaltered.
+
+  count may be uninitialized on input.  On return it is set to the
+  number of times the option was specified.  If the option is not
+  present, this is zero.  If an option has both a short and long name,
+  this is the sum of counts for both forms.
+
+  argument may be uninitialized on input.  On return it is set to the
+  last option-argument specified to this option, or NULL if no
+  option-argument was specified with the last occurrence of this
+  option, or the option was never specified.
+
+  flags is the bitwise-or of certain constants.  On input flags must
+  contain exactly one of:
+  GOPT_ARGUMENT_FORBIDDEN if the option does not take an option-argument
+  GOPT_ARGUMENT_REQUIRED if it always requires an option-argument
+  GOPT_ARGUMENT_OPTIONAL if it may or may not take an option-argument
+
+  If flags contains GOPT_ARGUMENT_REQUIRED, it may also contain
+  GOPT_ARGUMENT_NO_HYPHEN.  If the option argument is not allowed to
+  start with a '-', but the next member of the argument vector does
+  start with one where an option argument was expected, then this flag
+  causes it to be interpreted as another option, leaving the argument
+  to the preceding option NULL.  The application can then detect the
+  NULL and give an error message like "missing option argument" which
+  may be preferable to treating the next program argument as an option
+  argument and then giving a less helpful message when the option
+  argument is found to be invalid.
+
+  flags may also optionally contain GOPT_REPEATABLE.  This is ignored
+  by the option parsing code, but may be used when checking the value
+  returned in count, where if this flag is specified then the
+  application will presumably report an error if count is greater than
+  one.
+
+  On return, the initial value of flags is still present, and
+  additionally GOPT_SEEN_SHORT_WITH and/or GOPT_SEEN_SHORT_WITHOUT are
+  set if the short name was used at least once with or without an
+  option-argument, respectively, and GOPT_SEEN_LONG_WITH and/or
+  GOPT_SEEN_LONG_WITHOUT are set if the long name was used.
+
+  On input there must be zero or more elements in the options array
+  initialized as above.  These must be followed by one element where
+  flags is exactly GOPT_LAST.  All other members of this element may be
+  uninitialized on input.  On return the count member of this element
+  will be set to the total number of unrecognized options present.
+  The short_name member will be set to the first unrecognized short
+  option, or zero if there were none.  The long_name will be set to
+  point to the start of the name of the first unrecognized long option,
+  or NULL if there were none.  For completeness the argument member
+  will be set to the option argument of the last unrecognized option
+  (the same as with other options) but this is probably useless to the
+  application.
+*/
+
+void gopt_errors (const char *argv0, const struct option *options);
+/*
+  The function gopt_errors() examines the array of struct option
+  after it has been filled out by gopt().  If the options specified
+  were invalid it prints an error message starting with argv0 and
+  terminates the program.
+*/
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* GOPT_H_INCLUDED */

